core/metacling/src/TCling.cxx

```diff
@@ -5891,11 +5867,106 @@ Int_t TCling::AutoParse(const char *cls)
    return nHheadersParsed > 0 ? 1 : 0;
 }
 
+static void* LazyFunctionCreatorAutoloadForModule(const std::string& mangled_name,
+      cling::Interpreter *fInterpreter) {
+   using namespace llvm::object;
+   using namespace llvm::sys::path;
+   using namespace llvm::sys::fs;
+
+   static constexpr std::hash<std::string> hash_fn;
+   // size_t is a hashed mangled_name and libs[(uint8_t)mname_lib_table.second] = library's
+   // name in string.
+   static std::unordered_map<size_t, uint8_t> mname_lib_table;
+   static std::vector<std::string> libs;
+   // Initialize library table at first run
+   static bool fFirstRun = true;
+
+   R__LOCKGUARD(gInterpreterMutex);
+
+   // Initialize the table
+   if (fFirstRun) {
+      clang::Preprocessor &PP = fInterpreter->getCI()->getPreprocessor();
+      const HeaderSearchOptions &HSOpts = PP.getHeaderSearchInfo().getHeaderSearchOpts();
+      auto ModulePaths(HSOpts.PrebuiltModulePaths);
+
+      // sort and eliminate dupicate path
+      std::sort(ModulePaths.begin(), ModulePaths.end());
+      auto last = std::unique(ModulePaths.begin(), ModulePaths.end());
+      ModulePaths.erase(last, ModulePaths.end());
+
+      cling::DynamicLibraryManager* dyLibManager = fInterpreter->getDynamicLibraryManager();
+
+      int cur_lib = 0;
+      // Take path here eg. "/home/foo/module-release/lib/"
+      for (auto Path : ModulePaths) {
+         StringRef DirPath(Path);
+         if (DirPath.empty() || !is_directory(DirPath) || Path == ".")
+            continue;
+
+         // Check this not a random system directory (should contain pcm)
+         std::error_code EC;
+         bool is_build = false;
+         for (llvm::sys::fs::directory_iterator DirIt(DirPath, EC), DirEnd;
+               DirIt != DirEnd && !EC; DirIt.increment(EC)) {
+            std::string FileName(DirIt->path());
+            if (extension(FileName) == ".pcm") {
+               is_build = true;
+               break;
+            }
+         }
+
+         if (!is_build)
+            continue;
+
+         // Iterate over files under this path. We want to get each ".so" files
```

Comment:
Consider combining the two directory entry loops, creating a vector of library names while looking for the PCM file(s).